soccer/planning/planner/CollectPlanner.cpp

```diff
@@ -0,0 +1,139 @@
+            return control(std::move(request));
+    }
+    debugThrow("Error: CollectPlanner state machine failed.");
+    return Trajectory{{}};
+}
+Trajectory CollectPlanner::course(PlanRequest&& request) {
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    const Ball& ball = request.context->state.ball;
+    Point approachDir = findApproachDirection(request);
+    Point noisyTargetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius +
+                                  *_bufferDistBeforeContact);
+    Point noisyTargetVel =
+        approachDir * (ball.vel.dot(approachDir) + *_touchDeltaSpeed);
+    auto& targetInstant = _courseTargetInstants[request.shellID];
+    if (!targetInstant || targetInstant->pose.position().distTo(
+                              noisyTargetPoint) > *_targetChangeThreshold) {
+        targetInstant = RobotInstant{Pose{noisyTargetPoint, 0},
+                                     Twist{noisyTargetVel, 0}, RJ::Time{0s}};
+    }
+    request.motionCommand = PathTargetCommand{*targetInstant};
+    request.static_obstacles.add(
+        std::make_shared<Circle>(ball.pos, Ball_Radius));
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::fine(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    const Ball& ball = request.context->state.ball;
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    Point targetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius);
+    request.constraints.mot.maxSpeed =
+        ball.vel.dot(approachDir) + *_touchDeltaSpeed;
+    Point targetVel = request.constraints.mot.maxSpeed * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, Twist{targetVel, 0},
+                               RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::control(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    RobotInstant startInstant = request.start;
+    // move with constant negative acceleration until stop
+    double accel =
+        request.constraints.mot.maxAcceleration * *_controlAccelScalePercent;
+    double stoppingDist =
+        std::pow(startInstant.velocity.linear().mag(), 2) / (2 * accel);
+    Point targetPoint =
+        startInstant.pose.position() + stoppingDist * approachDir;
```

Comment:
You may need to add a config to extend this path slightly. If motion control is under-tuned (over-tuned?), it's possible this will not hit the ball due to stopping too soon.

Reply:
It also allowed for a few initial bounces on the ball off the dribbler and a way to tune to that